models/comment.rb

```diff
@@ -185,3 +204,63 @@ def set_sk
     logger.warn "NewRelic agent library not installed"
   end
 end
+
+
+def build_course_stats_for_user(user, course_id)
+  data = Content.collection.aggregate(
+    [
+      # Match all content in the course
+      { "$match" => { :course_id => course_id, :author_id => user.external_id } },
+      # Keep a count of flags for each entry
+      {
+        "$set" => {
+          # Just using $ne with null will return true if the field is absent
+          # So we first fall all absent fields to null, then check if it's null,
+          # that way we match for the absence of the field or value = null
+          :is_reply => { "$ne" => [{ "$ifNull" => ["$parent_id", nil] }, nil] }
+        }
+      },
+      {
+        "$group" => {
+          # Here we're grouping items by the type (comment or thread), and the user, and whether the comment is a reply.
+          # For threads is_reply will always be false.
+          :_id => { :type => "$_type", :is_reply => "$is_reply" },
+          # This will just count each group, so we get a breakdown of how many comments and threads a user has created.
+          :count => { "$sum" => 1 },
+          # These two will sum up the active and inactive reports in each category
+          # i.e. reported threads, reported comments, reported replies
+          # The way this works is (starting from inside out), we take the size of the abuse_flaggers list, we compare
+          # it to 0 using $cmp. If it's greater than zero then $cmp results in 1 otherwise 0.
+          # So we're summing up 1 for each abuse_flagger array that has entries, and 0 for the rest. This gives us a
+          # count of
+          :active_flags => { "$sum" => { "$cmp" => [{ "$size" => "$abuse_flaggers" }, 0] } },
+          :inactive_flags => { "$sum" => { "$cmp" => [{ "$size" => "$historical_abuse_flaggers" }, 0] } },
+        }
+      }
+    ])
```

Comment:
We use the same query, but scoped to a single user, so it's a much smaller set of data to go through.